cartography/intel/azure/cosmosdb.py

```diff
@@ -0,0 +1,1079 @@
+    SET d.lastupdated = {azure_update_tag},
+    d.kind = da.kind,
+    d.name = da.name,
+    d.ipranges = da.ipruleslist,
+    d.capabilities = da.list_of_capabilities,
+    d.documentendpoint = da.document_endpoint,
+    d.virtualnetworkfilterenabled = da.is_virtual_network_filter_enabled,
+    d.enableautomaticfailover = da.enable_automatic_failover,
+    d.provisioningstate = da.provisioning_state,
+    d.multiplewritelocations = da.enable_multiple_write_locations,
+    d.accountoffertype = da.database_account_offer_type,
+    d.publicnetworkaccess = da.public_network_access,
+    d.enablecassandraconnector = da.enable_cassandra_connector,
+    d.connectoroffer = da.connector_offer,
+    d.disablekeybasedmetadatawriteaccess = da.disable_key_based_metadata_write_access,
+    d.keyvaulturi = da.key_vault_key_uri,
+    d.enablefreetier = da.enable_free_tier,
+    d.enableanalyticalstorage = da.enable_analytical_storage,
+    d.defaultconsistencylevel = da.consistency_policy.default_consistency_level,
+    d.maxstalenessprefix = da.consistency_policy.max_staleness_prefix,
+    d.maxintervalinseconds = da.consistency_policy.max_interval_in_seconds
+    WITH d
+    MATCH (owner:AzureSubscription{id: {AZURE_SUBSCRIPTION_ID}})
+    MERGE (owner)-[r:RESOURCE]->(d)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {azure_update_tag}
+    """
+
+    for database_account in database_account_list:
+        capabilities: List[str] = []
+        iprules: List[str] = []
+        if 'capabilities' in database_account and len(database_account['capabilities']) > 0:
+            capabilities = [x['name'] for x in database_account['capabilities']]
+        if 'ip_rules' in database_account and len(database_account['ip_rules']) > 0:
+            iprules = [x['ip_address_or_range'] for x in database_account['ip_rules']]
+        database_account['ipruleslist'] = iprules
+        database_account['list_of_capabilities'] = capabilities
+
+    neo4j_session.run(
+        ingest_database_account,
+        database_accounts_list=database_account_list,
+        AZURE_SUBSCRIPTION_ID=subscription_id,
+        azure_update_tag=azure_update_tag,
+    )
+
+
+@timeit
+def load_database_account_data_resources(
+        neo4j_session: neo4j.Session, subscription_id: str, database_account_list: List[Dict], azure_update_tag: int,
+) -> None:
+    """
+    Loading all the resources (like cors policy, failover policy, private endpoint connections, virtual
+    network rules and locations) that we get as a part of the database account data itself.
+    """
+    for database_account in database_account_list:
+        # cleanup existing cors policy properties
+        run_cleanup_job(
+            'azure_cosmosdb_cors_details.json',
+            neo4j_session,
+            {'UPDATE_TAG': azure_update_tag, 'AZURE_SUBSCRIPTION_ID': subscription_id},
+        )
+
+        if 'cors' in database_account and len(database_account['cors']) > 0:
```

Comment:
Having all the data checks in this function makes it more complex and difficult to read. Can you move this logic to a `transform()` function so that the `load()` can work on processed objects? You might have seen this already but I'll link it here for others too: https://github.com/lyft/cartography/blob/master/docs/dev/writing-intel-modules.md#transform

Reply:
Alright! I have done so wherever applicable in the code :)

Reply:
This particular piece of code doesn't really transform anything. So I renamed it to `sync_database_account_data_resources()` since it just checks if the resources are present in the database account response (which we already have) and then runs their respective load functions if the resources are present.

That being said, I have added two transform functions in the code where the manipulation before neo4j ingestion was happening.

Reply:
This function does too many things at once. It does a transform+load for each type of policies/rules, and it does a transform+load for each type of location.

I'd recommend separating this for each resource.

I'll clarify this in a follow-up comment if needed; hold off on editing if you're  confused. It's getting late here but I wanted to make sure I gave you a somewhat timely second pass. :)

Reply:
@achantavy The response structure of the following resources - Locations, Cors Policies, Failover Policies, Private Endpoint Connections and Virtual Network Rules; comes inside the response structure of Database Accounts in the `get_database_account_list` function. 

The purpose of this function is to check if the Database Account response has the responses for the above mentioned resources and then call their respective load functions (if the resources are present). I'm not really transforming anything in the response structure here. Just a simple - "check if resource present -if yes, call load function; if no, move on".

I'll hold off on editing this one for now as I'm not fully understanding what you're trying to convey 🙈

Reply:
I agree with @achantavy. You should move the `if` blocks to their respective `_load_...` functions: If the list of things to load is empty, then don't run the query. Also if, for example, `database_account['failover_policies'])` exists, is it ever non-empty? If not, then you should do something like `if database_account.get('failover_policies'):`. 

And while you aren't really transforming the policies/rules, you certainly are doing a transformation against the `locations`.

Reply:
I moved the the `if` blocks to the respective `_load_` functions. And made some other major updates in this function by breaking it down into parts. Let me know if it is okay now.